fabtests/unit/getinfo_test.c

```diff
@@ -964,6 +964,477 @@ getinfo_test(caps, 4, "Test for capability bit regression",
+	hints->rx_attr->optional_caps = FI_MSG;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf,
+			       "fi_getinfo with optional_caps failed to return",
+			       ret);
+		return -FI_EINVAL;
+	}
+
+	ret = test_valid_cap_info(*info, required_caps,
+				  required_caps & (PRIMARY_TX_CAPS | OPT_TX_CAPS),
+				  required_caps & (PRIMARY_RX_CAPS | OPT_RX_CAPS));
+	if (ret) {
+		FT_UNIT_STRERR(err_buf,
+			       "Required caps, set via optional_caps, missing",
+			       ret);
+		return ret;
+	}
+
+	ret = test_valid_optional_cap_info(*info, 0, 0, 0);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "Optional caps incorrectly returned",
+			       ret);
+		return ret;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int test_valid_suboptimal_optional_cap(char *node, char *service,
+					      uint64_t flags,
+					      struct fi_info *hints,
+					      struct fi_info **info)
+{
+	uint64_t required_caps = FI_ATOMIC | FI_TAGGED;
+	int ret;
+
+	/* Verify that the provider meets the required capabilities of this
+	 * test.
+	 */
+	hints->caps = required_caps;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		sprintf(err_buf, "No providers found");
+		return -FI_ENOSYS;
+	}
+
+	/* This test is only setup to support selected providers. */
+	if (!strstr(hints->fabric_attr->prov_name, "ofi_rxm")) {
+		sprintf(err_buf, "%s not support for this test",
+			hints->fabric_attr->prov_name);
+		return -FI_ENOSYS;
+	}
+
+	/* Since fi_getinfo() may return secondary caps, need to update
+	 * required_caps.
+	 */
+	required_caps = (*info)->caps;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Pass in the majority of the require test capabilities as optional.
+	 * This capabilites should be set in the return fi_info::caps. In
+	 * addition, since user hint domain_attr::caps, tx_attr::caps, and
+	 * rx_attr::caps are zero, these returns caps should be set based on
+	 * fi_info::caps
+	 */
+	hints->caps = FI_TAGGED;
+	hints->optional_caps = FI_ATOMIC;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_UNIT_STRERR(err_buf,
+			       "fi_getinfo with optional_caps failed to return",
+			       ret);
+		return -FI_EINVAL;
+	}
+
+	ret = test_valid_optional_cap_info(*info, FI_ATOMIC, FI_ATOMIC,
+					   FI_ATOMIC);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "Optional caps incorrectly returned",
+			       ret);
+		return ret;
+	}
+
+	/* For RxM, FI_ATOMIC used as optional_cap fi_getinfo() input should not
+	 * be returned in caps output since performance could be impacted.
+	 */
+	ret = test_invalid_cap_info(*info, FI_ATOMIC, FI_ATOMIC, FI_ATOMIC);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf,
+			       "FI_ATOMIC incorrectly returned in caps",
+			       ret);
+		return ret;
+	}
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Verify that using domain_attr::optional_caps, tx_attr::optional_caps,
```

Comment:
I'm not following the purpose of the testing in this function from here down.

Reply:
This test is trying to verify the hints::{tx,rx}_attr::optional_caps against a super-set of hints::caps.

1. With all optional_caps == 0, verify that only the user request hints::{tx,rx}_attr::caps, which are a subset of hints::caps, is returned in the fi_info from fi_getinfo(). This is the default behavior of fi_getinfo() and really doesn't test optional capabilities.
1. Adding in hints::{tx,rx}_attr::optional_caps == FI_ATOMIC to the previous test, the returned fi_info::{tx,rx}_attr::optional_caps (not caps) should now contain FI_ATOMIC. For RxM, since FI_ATOMIC may negatively impact performance, when requested as a optional_cap, it should only be returned in optional_cap output.